File: platforms/windows/Zylix/TodoViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;

namespace Zylix;

/// <summary>
/// Represents a single todo item
/// </summary>
public class TodoItem : INotifyPropertyChanged
{
    private bool _isCompleted;

    public uint Id { get; init; }
    public string Text { get; init; } = string.Empty;

    public bool IsCompleted
    {
        get => _isCompleted;
        set
        {
            if (_isCompleted != value)
            {
                _isCompleted = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsCompleted)));
            }
        }
    }

    public event PropertyChangedEventHandler? PropertyChanged;
}

/// <summary>
/// Filter mode for todos
/// </summary>
public enum TodoFilter
{
    All,
    Active,
    Completed
}

/// <summary>
/// ViewModel for the Todo app
/// </summary>
public class TodoViewModel : INotifyPropertyChanged
{
    private string _newTodoText = string.Empty;
    private TodoFilter _currentFilter = TodoFilter.All;
    private uint _nextId = 1;
    private int _renderCount;
    private double _lastRenderTime;
    private readonly Stopwatch _stopwatch = new();

    public ObservableCollection<TodoItem> Items { get; } = new();

    public string NewTodoText
    {
        get => _newTodoText;
        set
        {
            if (_newTodoText != value)
            {
                _newTodoText = value;
                OnPropertyChanged(nameof(NewTodoText));
                OnPropertyChanged(nameof(CanAddTodo));
            }
        }
    }

    public TodoFilter CurrentFilter
    {
        get => _currentFilter;
        set
        {
            if (_currentFilter != value)
            {
                _currentFilter = value;
                OnPropertyChanged(nameof(CurrentFilter));
                OnPropertyChanged(nameof(FilteredItems));
                OnPropertyChanged(nameof(IsAllFilter));
                OnPropertyChanged(nameof(IsActiveFilter));
                OnPropertyChanged(nameof(IsCompletedFilter));
            }
        }
    }

    public bool IsAllFilter => CurrentFilter == TodoFilter.All;
    public bool IsActiveFilter => CurrentFilter == TodoFilter.Active;
    public bool IsCompletedFilter => CurrentFilter == TodoFilter.Completed;

    public ObservableCollection<TodoItem> FilteredItems
    {
        get
        {
            var filtered = CurrentFilter switch
            {
                TodoFilter.Active => Items.Where(x => !x.IsCompleted),
                TodoFilter.Completed => Items.Where(x => x.IsCompleted),
                _ => Items
            };
            return new ObservableCollection<TodoItem>(filtered);
        }
    }

    public int ActiveCount => Items.Count(x => !x.IsCompleted);
    public int CompletedCount => Items.Count(x => x.IsCompleted);
    public bool AllCompleted => Items.Count > 0 && Items.All(x => x.IsCompleted);
    public bool HasCompleted => CompletedCount > 0;
    public bool CanAddTodo => !string.IsNullOrWhiteSpace(NewTodoText);

    public int RenderCount
    {
        get => _renderCount;
        private set
        {
            _renderCount = value;
            OnPropertyChanged(nameof(RenderCount));
            OnPropertyChanged(nameof(StatsText));
        }
    }

    public double LastRenderTime
    {
        get => _lastRenderTime;
        private set
        {
            _lastRenderTime = value;
            OnPropertyChanged(nameof(LastRenderTime));
            OnPropertyChanged(nameof(StatsText));
        }
    }

    public string StatsText => $"{Items.Count} Todos  |  {RenderCount} Renders  |  {LastRenderTime:F2} ms";
    public string ItemsLeftText => $"{ActiveCount} item{(ActiveCount == 1 ? "" : "s")} left";

    public event PropertyChangedEventHandler? PropertyChanged;

    public TodoViewModel()
    {
        // Add sample todos
        AddTodo("Learn Zig");
        AddTodo("Build VDOM");
        AddTodo("Create Windows bindings");
    }

    public void AddTodo(string? text = null)
    {
        var todoText = text ?? NewTodoText.Trim();
        if (string.IsNullOrWhiteSpace(todoText)) return;

        _stopwatch.Restart();

        var item = new TodoItem
        {
            Id = _nextId++,
            Text = todoText,
            IsCompleted = false
        };
        Items.Add(item);

        if (text == null)
        {
            NewTodoText = string.Empty;
        }

        TrackRender();
        NotifyAllChanged();
    }

    public void RemoveTodo(TodoItem item)
    {
        _stopwatch.Restart();
        Items.Remove(item);
        TrackRender();
        NotifyAllChanged();
    }

    public void ToggleTodo(TodoItem item)
    {
        _stopwatch.Restart();
        item.IsCompleted = !item.IsCompleted;
        TrackRender();
        NotifyAllChanged();
    }

    public void ToggleAll()
    {
        _stopwatch.Restart();
        var shouldComplete = !AllCompleted;
        foreach (var item in Items)
        {
            item.IsCompleted = shouldComplete;
        }
        TrackRender();
        NotifyAllChanged();
    }

    public void ClearCompleted()
    {
        _stopwatch.Restart();
        var completed = Items.Where(x => x.IsCompleted).ToList();
        foreach (var item in completed)
        {
            Items.Remove(item);
        }
        TrackRender();
        NotifyAllChanged();
    }

    public void SetFilter(TodoFilter filter)
    {
        CurrentFilter = filter;
    }

    private void TrackRender()
    {
        _stopwatch.Stop();
        LastRenderTime = _stopwatch.Elapsed.TotalMilliseconds;
        RenderCount++;
    }

    private void NotifyAllChanged()
    {
        OnPropertyChanged(nameof(FilteredItems));
        OnPropertyChanged(nameof(ActiveCount));
        OnPropertyChanged(nameof(CompletedCount));
        OnPropertyChanged(nameof(AllCompleted));
        OnPropertyChanged(nameof(HasCompleted));
        OnPropertyChanged(nameof(ItemsLeftText));
        OnPropertyChanged(nameof(StatsText));
    }

    private void OnPropertyChanged(string propertyName)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}

File: platforms/ios/Zylix/Sources/TodoViewModel.swift
//
//  TodoViewModel.swift
//  Zylix
//
//  Todo state management using pure Swift.
//  Future: Integrate with Zylix Core via C FFI.
//

import Foundation
import SwiftUI

// MARK: - Todo Item

struct TodoItem: Identifiable, Equatable {
    let id: UUID
    var text: String
    var isCompleted: Bool

    init(text: String, isCompleted: Bool = false) {
        self.id = UUID()
        self.text = text
        self.isCompleted = isCompleted
    }
}

// MARK: - Filter Mode

enum FilterMode: String, CaseIterable {
    case all = "All"
    case active = "Active"
    case completed = "Completed"
}

// MARK: - Todo View Model

@MainActor
final class TodoViewModel: ObservableObject {
    @Published var items: [TodoItem] = []
    @Published var filter: FilterMode = .all
    @Published var newTodoText: String = ""
    @Published var editingItem: TodoItem?

    // MARK: - Computed Properties

    var filteredItems: [TodoItem] {
        switch filter {
        case .all:
            return items
        case .active:
            return items.filter { !$0.isCompleted }
        case .completed:
            return items.filter { $0.isCompleted }
        }
    }

    var activeCount: Int {
        items.filter { !$0.isCompleted }.count
    }

    var completedCount: Int {
        items.filter { $0.isCompleted }.count
    }

    var allCompleted: Bool {
        !items.isEmpty && items.allSatisfy { $0.isCompleted }
    }

    var itemsLeftText: String {
        let count = activeCount
        return count == 1 ? "1 item left" : "\(count) items left"
    }

    // MARK: - Actions

    func addTodo() {
        let trimmed = newTodoText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }

        let item = TodoItem(text: trimmed)
        items.append(item)
        newTodoText = ""
    }

    func toggle(_ item: TodoItem) {
        guard let index = items.firstIndex(where: { $0.id == item.id }) else { return }
        items[index].isCompleted.toggle()
    }

    func remove(_ item: TodoItem) {
        items.removeAll { $0.id == item.id }
    }

    func toggleAll() {
        let newState = !allCompleted
        for index in items.indices {
            items[index].isCompleted = newState
        }
    }

    func clearCompleted() {
        items.removeAll { $0.isCompleted }
    }

    func updateText(_ item: TodoItem, newText: String) {
        guard let index = items.firstIndex(where: { $0.id == item.id }) else { return }
        let trimmed = newText.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            items.remove(at: index)
        } else {
            items[index].text = trimmed
        }
    }
}
